rfcs/0020-telemetry.md

```diff
@@ -0,0 +1,192 @@
+- Feature Name: Telemetry
+
+- Start Date: 2024-01-24
+
+- RFC PR: [mbta/technology-docs#0000](https://github.com/mbta/technology-docs/pull/0000)
+
+- Asana task: [asana link](https://app.asana.com/)
+
+- Status: Proposed
+
+# Summary
+
+Observability is the ability to measure the internal states of a system by examining its outputs.<sup>[1](https://www.splunk.com/en_us/blog/learn/observability.html)</sup>
+It's what allows us to understand how our applications and systems work.
+There are three pillars to observability:<sup>[2](https://www.crowdstrike.com/cybersecurity-101/observability/three-pillars-of-observability/)</sup>
+
+  1. **Logs** - are the archival or historical records of system events and errors, which can be plain text, binary, or structured with metadata.
+  2. **Metrics** - are numerical measurements of system performance and behavior, such as CPU usage, response time, or error rate.
+  3. **Traces** - are the representations of individual requests or transactions that flow through a system, which can help identify bottlenecks, dependencies, and root causes of issues.
+
+TID engineers currently make extensive use of logging.
+But, we don't currently have an established mechanism for collecting metrics or traces.
+This RFC proposes multiple pathways for collecting metrics in our current observability platform, Splunk.
+
+# Motivation
+
+Logs are handy for debugging and other unstructured research like a security audit.
+However, many TID developers have noted that some queries based on logs are incredibly slow, if not impossible, to run.
+
+> It might be tempting to think that logs can solve every use case. As the amount of data grows, however, a logs-only solution will become costly and relatively slow for a small set of regular searches, usually connected to alerts. This is because the process by which logs must be categorized and batched takes much more time and is much more computationally intensive than the metrics process...<sup>[3](https://www.splunk.com/en_us/blog/learn/logs-vs-metrics.html)</sup>
+
+But, as noted above, logs are just one of three tools used to observe systems.
+Metrics are optimized for storage and retrieval as numbers in a time series, which makes them cheap to store and fast to retrieve.
+
+By adopting metrics, we can speed up slow queries, decrease the time necessary to generate dashboards and reports, and make previously impossible alerts possible.
```

Comment:
Yeah, let's say you want to know if the number of requests you received over the last minute was anomalous. You would want to compare that minute to other bins over the past week. Looking at logs over the past week takes too long...because they aren't optimized to be viewed that way.

Reply:
We have a alert which does this for another application already: search for "Concentrate: Unusual Metric Drop" in the Splunk alerts.